File: src/main/java/om/example/om_pay/model/Utilisateur.java
package om.example.om_pay.model;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import om.example.om_pay.model.enums.Role;
import om.example.om_pay.model.enums.Statut;

@Entity
@Table(name = "utilisateur")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Utilisateur {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String nom;
    private String prenom;

    @Column(unique = true, nullable = false)
    private String telephone;

    private String email;

    @Column(length = 255, nullable = false)
    private String motDePasse; 

    @Column(length = 255, nullable = true)
    private String codePin;  

    @Enumerated(EnumType.STRING)
    private Role role;  

    @Enumerated(EnumType.STRING)
    private Statut statut; 

    private Double plafondQuotidien = 500000.0;  
    private Double totalTransfertJour = 0.0;     
    private LocalDate dernierResetPlafond;      

    @Column(length = 6)
    private String codeSecret;  
    
    private LocalDateTime codeSecretExpiration;  
    
    private Boolean premiereConnexion = true;  

    private LocalDateTime dateCreation = LocalDateTime.now();
    private LocalDateTime dateModification;

    @OneToMany(mappedBy = "utilisateur", cascade = CascadeType.ALL)
    private List<Compte> comptes;

    @OneToMany(mappedBy = "distributeur")
    private List<Transaction> operationsDistributeur;
}

File: apprendre_dart/lib/entitties/utilisateur.dart
import '../enums/role.dart';
import '../enums/statut.dart';

class Utilisateur {
  final int? id;
  final String nom;
  final String prenom;
  final String telephone;
  final String email;
  final String? motDePasse;
  final Role role;
  final Statut statut;
  final bool premiereConnexion;
  final String dateCreation;

  Utilisateur({
    this.id,
    required this.nom,
    required this.prenom,
    required this.telephone,
    required this.email,
    this.motDePasse,
    required this.role,
    required this.statut,
    this.premiereConnexion = true,
    required this.dateCreation,
  });

  factory Utilisateur.fromJson(Map<String, dynamic> json) {
    return Utilisateur(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id'].toString()),
      nom: json['nom'],
      prenom: json['prenom'],
      telephone: json['telephone'],
      email: json['email'],
      motDePasse: json['motDePasse'],
      role: Role.fromString(json['role']),
      statut: Statut.fromString(json['statut']),
      premiereConnexion: json['premiereConnexion'] ?? true,
      dateCreation: json['dateCreation'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'nom': nom,
      'prenom': prenom,
      'telephone': telephone,
      'email': email,
      if (motDePasse != null) 'motDePasse': motDePasse,
      'role': role.name,
      'statut': statut.name,
      'premiereConnexion': premiereConnexion,
      'dateCreation': dateCreation,
    };
  }

  @override
  String toString() {
    return 'Utilisateur(id: $id, nom: $nom, prenom: $prenom, email: $email, role: $role)';
  }
}